[ao] Fix per-channel histogram visualization in ModelReportVisualizer

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82918
* __->__ #82917

Summary: There was an issue with per-channel visualizations in the
ModelReportVisualizer that in specific scenarios in which there were
only per-channel features for a module, it would fail to specifically
get the channel by channel info.

After digging through the code, the core reason was a for loop that was
enumerating on the `tensor_table` (tensor level info) even in the
scenario in which we only had per-channel info.

This was fixed, and tested in a Bento to ensure expected functionality.

Test Plan: Tested visually

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Fix per-channel histogram visualization in ModelReportVisualizer

Summary: There was an issue with per-channel visualizations in the
ModelReportVisualizer that in specific scenarios in which there were
only per-channel features for a module, it would fail to specifically
get the channel by channel info.

After digging through the code, the core reason was a for loop that was
enumerating on the `tensor_table` (tensor level info) even in the
scenario in which we only had per-channel info.

This was fixed, and tested in a Bento to ensure expected functionality.

Test Plan: Tested visually

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -486,25 +487,26 @@ def _get_plottable_data(self, feature_filter: str, module_fqn_filter: str) -> Tu
                 # gather the x_data and multiple y_data
                 # calculate the number of channels
-                num_channels: int = max(row[self.CHANNEL_NUM_INDEX] for row in channel_table) + 1
+                num_channels: int = max(row[self.CHANNEL_NUM_INDEX] for row in table) + 1
```
I feel ideally we should just split these two cases, it might be easier to maintain

vspenubarthi replied:
i actually originally had it as two different cases in an earlier PR and the feedback i got was to merge the code together as much as possible to reduce redundancy. 

I think the main argument against two different cases is that the method will get long and there will be a good amount of redundant code.

jerryzh168 replied:
I feel it's fine to repeat if you can't actually merge them, in this case, the use of `if is_valid_per_channel_plot` is revealing that they can't be truely merged, is it possible to not use this `if is_valid_per_channel_plot`?

vspenubarthi replied:
i tried without it and couldn't get it to work as desired. I split up into distinct sections

vspenubarthi committed:
Update on "[ao] Fix per-channel histogram visualization in ModelReportVisualizer"


Summary: There was an issue with per-channel visualizations in the
ModelReportVisualizer that in specific scenarios in which there were
only per-channel features for a module, it would fail to specifically
get the channel by channel info.

After digging through the code, the core reason was a for loop that was
enumerating on the `tensor_table` (tensor level info) even in the
scenario in which we only had per-channel info.

This was fixed, and tested in a Bento to ensure expected functionality.

Test Plan: Tested visually

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
looks good, thanks

vspenubarthi commented:
@pytorchbot merge -g
